frontend/app/src/App.tsx

```diff
@@ -1201,27 +1201,31 @@ export class App extends PureComponent<Props, State> {
         this.state.scriptFinishedHandlers.map(handler => handler())
       }, 0)
 
-      // Clear any stale elements left over from the previous run.
-      // (We don't do this if our script had a compilation error and didn't
-      // finish successfully.)
-      this.setState(
-        ({ scriptRunId, fragmentIdsThisRun }) => ({
-          // Apply any pending elements that haven't been applied.
-          elements: this.pendingElementsBuffer.clearStaleNodes(
-            scriptRunId,
-            fragmentIdsThisRun
-          ),
-        }),
-        () => {
-          this.pendingElementsBuffer = this.state.elements
-        }
-      )
-
       if (
         status === ForwardMsg.ScriptFinishedStatus.FINISHED_SUCCESSFULLY ||
         status ===
           ForwardMsg.ScriptFinishedStatus.FINISHED_FRAGMENT_RUN_SUCCESSFULLY
       ) {
+        // Clear any stale elements left over from the previous run.
+        // We only do that for completed runs, not for runs that were finished early
+        // due to reruns; this is to avoid flickering of elements where they disappear for
```

Comment:
nit: what type of finished early rerun does it refer to: e.g. also one triggered by `st.rerun`?

Reply:
> I'm wondering if there is any way to implement some e2e tests around stale element handling. It has been causing some trouble and feels like to fully covered by tests yet. Doesn't have to be part of this PR, but could be a dedicated project to add some e2e testing for stale handling.

Yeah that's in general a good point. Most of these bugs are difficult to test in e2e tests because they only appear sometimes and usually are caused by race conditions which can easily lead to flaky tests. So usually unit tests seem to be better for these tests. There are some tests that check for cleaning up stale widgets properly, e.g. the `st_fragments_nested` e2e playwright test.

Reply:
We probably can add a bit more e2e testing around aspects like:
- One element is exchanged with a different element. 
- An element is added in between different element
- There is a computation with a long load time during the script execution.
- Usage of placeholder elements.
- ... 
with all these scenarios, we would need to make sure that only the expected elements are on screen / shown as stale. But it is a bit more tricky to get that implemented in a non-flaky way. I added a tech debt entry [here](https://www.notion.so/snowflake-corp/Add-e2e-test-around-execution-flow-and-stale-handling-43c222bdd519443eae5b98b6450e5035).

Reply:
Thank you!